go/src/koding/klientctl/ssh/sshold.go

```diff
@@ -0,0 +1,408 @@
+// machine. On subsequent requests it uses the same key pair, but adds it each
+// time to user machine, since we can't assume if key exists in local, remote
+// machine must also have that key, ie if user has multiple machines, but key
+// was only added to one machine.
+//
+// Before generating a new key, it asks user to confirm or deny.
+//
+// SSH public keys have comment of the form: "koding-<number>", where number is
+// a random integer. Klient requires a non empty comment; by adding a random
+// integer to end of comment this command can be used by multiple computers.
+type SSHCommand struct {
+	*SSHKey
+
+	Log   logging.Logger
+	Debug bool
+
+	// Ask is flag for user interaction, ie should we ask user to generate new
+	// SSH key if it doesn't exist.
+	Ask bool
+
+	// Klient is communication layer between this and local klient. This is
+	// used to add SSH public key to `~/.ssh/authorized_keys` on the remote
+	// machine.
+	Klient interface {
+		RemoteList() (list.KiteInfos, error)
+	}
+}
+
+// NewSSHCommand is the required initializer for SSHCommand.
+func NewSSHCommand(log logging.Logger, opts SSHCommandOpts) (*SSHCommand, error) {
+	usr, err := user.Current()
+	if err != nil {
+		return nil, err
+	}
+
+	klientKite, err := klient.CreateKlientWithDefaultOpts()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := klientKite.Dial(); err != nil {
+		log.New("NewSSHCommand").Error("Dialing local klient failed. err:%s", err)
+		return nil, ErrLocalDialingFailed
+	}
+
+	k := klient.NewKlient(klientKite)
+
+	return &SSHCommand{
+		Klient: k,
+		Log:    log.New("SSHCommand"),
+		Ask:    opts.Ask,
+		Debug:  opts.Debug,
+		SSHKey: &SSHKey{
+			Log:            log.New("SSHKey"),
+			Debug:          opts.Debug,
+			RemoteUsername: opts.RemoteUsername,
+			KeyPath:        path.Join(usr.HomeDir, config.SSHDefaultKeyDir),
+			KeyName:        config.SSHDefaultKeyName,
+			Klient:         k,
+		},
+	}, nil
+}
+
+func (s *SSHCommand) Run(machine string) error {
+	if !s.KeysExist() && s.Ask {
+		util.MustConfirm("'ssh' command needs to create public/private rsa key pair. Continue? [Y|n]")
+	}
+
+	machine, err := shortcut.NewMachineShortcut(s.Klient).GetNameFromShortcut(machine)
+	if err != nil {
+		return err
+	}
+
+	userhost, port, err := s.GetSSHAddr(machine)
+	if err != nil {
+		return err
+	}
+
+	if err := s.PrepareForSSH(machine); err != nil {
+		s.Log.Debug("PrepareForSSH returned err: %s", err)
+
+		if strings.Contains(err.Error(), "user: unknown user") {
+			return ErrCannotFindUser
+		}
+
+		// TODO: We're unable to log the meaningful error returned from klient, so we're
+		// leaking possibly meaningful data here. This will be resolved once SSH gets
+		// updated to the new (final) format. Fix this.
+		if klientctlerrors.IsMachineNotValidYetErr(err) {
+			return ErrMachineNotValidYet
+		}
+
+		if klientctlerrors.IsDialFailedErr(err) {
+			return ErrRemoteDialingFailed
+		}
+
+		return ErrFailedToGetSSHKey
+	}
+
+	args := []string{
```

Comment:
Can we add here something like `-o StrictHostKeychecking=no` and `-o UserKnownHostsFile=/dev/null`, in case of ephemeral IPs vms can re-use a host that was already added, which makes the ssh command to fail. Happened to me a number of times.